### ISSUE TITLE
RuntimeError when saving PDFs via parallel processes (not threads!)

### ISSUE THREAD

**Issue by marcotama** [matplotlib/matplotlib#13723]
### Bug report

**Bug summary**

When storing multiple PDFs using parallel processes (not threads!) a RuntimeError occurs. This happens regardless of the library used for parallelization, i.e. `concurrent`, `multiprocessing` or `joblib`. The error is either `RuntimeError: In set_text: could not load glyph` (generated by the code below) or `RuntimeError: In load_char: Could not load charcode` (which I can't seem to reproduce yet, but maybe you have a better idea of the link behind the two).

**Code for reproduction**

```python
from multiprocessing import Pool
from joblib import Parallel, delayed
from concurrent.futures import ThreadPoolExecutor, ProcessPoolExecutor, as_completed

def model_handler(n):
    import matplotlib.pyplot as plt
    fig_out = plt.figure(figsize=(1, 1))
    axes = fig_out.add_subplot(1,1,1)
    axes.scatter(range(1000), range(1000))
    fig_out.savefig('dummyfig{}.pdf'.format(n), format='pdf')
    plt.close()
    return n

n_jobs = 2
n_workers = 2

print('List comprehension (serial)...                           ', end='')
try:
    [model_handler(j) for j in range(n_jobs)]
    print('SUCCESS')
except:
    print('FAILURE')

print('Parallel processes via multiprocessing package...        ', end='')
try:
    Pool(processes=n_workers).map(model_handler, range(n_jobs))
    print('SUCCESS')
except:
    print('FAILURE')

print('Parallel processes via joblib package (loky backend)...  ', end='')
try:
    Parallel(n_jobs=n_workers, backend='loky')(delayed(model_handler)(i) for i in range(n_jobs))
    print('SUCCESS')
except:
    print('FAILURE')

print('Parallel threads via joblib package...                   ', end='')
try:
    Parallel(n_jobs=n_workers, backend='threading')(delayed(model_handler)(i) for i in range(n_jobs))
    print('SUCCESS')
except:
    print('FAILURE')

print('Parallel processes via joblib package...                 ', end='')
try:
    Parallel(n_jobs=n_workers, backend='multiprocessing')(delayed(model_handler)(i) for i in range(n_jobs))
    print('SUCCESS')
except:
    print('FAILURE')

print('Parallel threads via concurrent.futures package...       ', end='')
with ThreadPoolExecutor(max_workers=n_workers) as pool:
    futures = [pool.submit(model_handler, i) for i in range(n_jobs)]
    exceptions = [x.exception() for x in as_completed(futures)]
if all(exception is None for exception in exceptions):
    print('SUCCESS')
else:
    print('FAILURE')

print('Parallel processes via concurrent.futures package...     ', end='')
with ProcessPoolExecutor(max_workers=n_workers) as pool:
    futures = [pool.submit(model_handler, i) for i in range(n_jobs)]
    exceptions = [x.exception() for x in as_completed(futures)]
if all(exception is None for exception in exceptions):
    print('SUCCESS')
else:
    print('FAILURE')
```

**Actual outcome**

```
List comprehension (serial)...                           SUCCESS
Parallel processes via multiprocessing package...        FAILURE
Parallel processes via joblib package (loky backend)...  SUCCESS
Parallel threads via joblib package...                   SUCCESS
Parallel processes via joblib package...                 FAILURE
Parallel threads via concurrent.futures package...       SUCCESS
Parallel processes via concurrent.futures package...     FAILURE
```

**Traceback**

```
"""
Traceback (most recent call last):
  File "/Users/marcotama/anaconda3/envs/py37/lib/python3.7/multiprocessing/pool.py", line 121, in worker
    result = (True, func(*args, **kwds))
  File "/Users/marcotama/anaconda3/envs/py37/lib/python3.7/multiprocessing/pool.py", line 44, in mapstar
    return list(map(*args))
  File "<ipython-input-16-c44ed7855af1>", line 13, in model_handler
    fig_out.savefig('dummyfig{}.pdf'.format(n), format='pdf')
  File "/Users/marcotama/anaconda3/envs/py37/lib/python3.7/site-packages/matplotlib/figure.py", line 2035, in savefig
    self.canvas.print_figure(fname, **kwargs)
  File "/Users/marcotama/anaconda3/envs/py37/lib/python3.7/site-packages/matplotlib/backend_bases.py", line 2263, in print_figure
    **kwargs)
  File "/Users/marcotama/anaconda3/envs/py37/lib/python3.7/site-packages/matplotlib/backends/backend_pdf.py", line 2586, in print_pdf
    self.figure.draw(renderer)
  File "/Users/marcotama/anaconda3/envs/py37/lib/python3.7/site-packages/matplotlib/artist.py", line 55, in draw_wrapper
    return draw(artist, renderer, *args, **kwargs)
  File "/Users/marcotama/anaconda3/envs/py37/lib/python3.7/site-packages/matplotlib/figure.py", line 1475, in draw
    renderer, self, artists, self.suppressComposite)
  File "/Users/marcotama/anaconda3/envs/py37/lib/python3.7/site-packages/matplotlib/image.py", line 141, in _draw_list_compositing_images
    a.draw(renderer)
  File "/Users/marcotama/anaconda3/envs/py37/lib/python3.7/site-packages/matplotlib/artist.py", line 55, in draw_wrapper
    return draw(artist, renderer, *args, **kwargs)
  File "/Users/marcotama/anaconda3/envs/py37/lib/python3.7/site-packages/matplotlib/axes/_base.py", line 2607, in draw
    mimage._draw_list_compositing_images(renderer, self, artists)
  File "/Users/marcotama/anaconda3/envs/py37/lib/python3.7/site-packages/matplotlib/image.py", line 141, in _draw_list_compositing_images
    a.draw(renderer)
  File "/Users/marcotama/anaconda3/envs/py37/lib/python3.7/site-packages/matplotlib/artist.py", line 55, in draw_wrapper
    return draw(artist, renderer, *args, **kwargs)
  File "/Users/marcotama/anaconda3/envs/py37/lib/python3.7/site-packages/matplotlib/axis.py", line 1192, in draw
    renderer)
  File "/Users/marcotama/anaconda3/envs/py37/lib/python3.7/site-packages/matplotlib/axis.py", line 1130, in _get_tick_bboxes
    extent = tick.label1.get_window_extent(renderer)
  File "/Users/marcotama/anaconda3/envs/py37/lib/python3.7/site-packages/matplotlib/text.py", line 922, in get_window_extent
    bbox, info, descent = self._get_layout(self._renderer)
  File "/Users/marcotama/anaconda3/envs/py37/lib/python3.7/site-packages/matplotlib/text.py", line 300, in _get_layout
    ismath=False)
  File "/Users/marcotama/anaconda3/envs/py37/lib/python3.7/site-packages/matplotlib/backends/backend_pdf.py", line 2164, in get_text_width_height_descent
    font.set_text(s, 0.0, flags=LOAD_NO_HINTING)
RuntimeError: In set_text: could not load glyph
"""
The above exception was the direct cause of the following exception:
Traceback (most recent call last):
  File "/Users/marcotama/anaconda3/envs/py37/lib/python3.7/site-packages/IPython/core/interactiveshell.py", line 3267, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-18-426f53ccd8fe>", line 3, in <module>
    Pool(processes=n_workers).map(model_handler, range(n_jobs))
  File "/Users/marcotama/anaconda3/envs/py37/lib/python3.7/multiprocessing/pool.py", line 268, in map
    return self._map_async(func, iterable, mapstar, chunksize).get()
  File "/Users/marcotama/anaconda3/envs/py37/lib/python3.7/multiprocessing/pool.py", line 657, in get
    raise self._value
RuntimeError: In set_text: could not load glyph
```

**Expected outcome**

The code should simply succeed. Notice that the `loky` backend of `joblib` succeeds even though it uses parallel processes. I am not sure why this is.

**Matplotlib version**

  * Operating system: MacOS High Sierra (10.13.6)
  * Matplotlib version: 2.2.2
  * Matplotlib backend (`print(matplotlib.get_backend())`): `module://backend_interagg`
  * Python version: 3.7.2
  * Jupyter version (if applicable): N/A
  * Other libraries: joblib==0.12.2

All packages were installed via `pip` inside a `conda` environment.

**Comment by marcotama** [comment]
I found only one other reference to the above errors, here: https://github.com/uqfoundation/pathos/issues/153

**Comment by jklymak** [comment]
Did you try `plt.close(fig_out)` just on the hunch that pyplot is losing track of the current figure?

**Comment by marcotama** [comment]
Good idea. I tried that just now: unfortunately the problem persists. I think that only applies to the threading case, as separate processes run on whole separate instances of Python.

**Comment by anntzer** [comment]
I'm not sure what's happening here, but interestingly enough, the bug does not occur with https://github.com/anntzer/mplcairo :)

**Comment by anntzer** [comment]
Looks like adding
```
if __name__ == "__main__":
    import multiprocessing as mp; mp.set_start_method("spawn")
```
at the top of everything and additionally moving all the pool-spawning under `if __name__ == "__main__"` fixes the problem (at least for me).

I guess ft2font doesn't like being forked...

**Comment by marcotama** [comment]
If I run the above (with or without the "spawn" edit) on the command line (instead of the Python REPL as I did initially) I get a bunch of

```
The process has forked and you cannot use this CoreFoundation functionality safely. You MUST exec().
Break on __THE_PROCESS_HAS_FORKED_AND_YOU_CANNOT_USE_THIS_COREFOUNDATION_FUNCTIONALITY___YOU_MUST_EXEC__() to debug.
```

A [StackOverflow question](https://stackoverflow.com/questions/15871650/the-process-has-forked-and-you-cannot-use-this-corefoundation-functionality) cites [Apple's Obsolete Programming Techniques](https://developer.apple.com/library/archive/technotes/tn2083/_index.html#//apple_ref/doc/uid/DTS10003794-CH1-SUBSECTION64) which states

> Many Mac OS X frameworks do not work reliably if you call fork but do not call exec. The only exception is the System framework and, even there, the POSIX standard places severe constraints on what you can do between a fork and an exec.
> Under the covers the daemon routine calls fork. Thus, if you call daemon and use frameworks above the System frameworks, you may run into problems. Critically, frameworks that are otherwise daemon safe, like Core Foundation, are not safe to call from this state.

and 

> Important: In fact, in Mac OS X 10.5 and later, Core Foundation will detect this situation and print the warning message shown in Listing 13

(Listing 13 contains the error message I got)

Could the above be the root of the issue? Is there anything to do about it?

**Comment by marcotama** [comment]
I should mention, the code terminates with a segmentation fault.

**Comment by saravanabalagi** [comment]
I can confirm that this is not just with PDFs, I got it saving SVGs

**Comment by saravanabalagi** [comment]
Is there any temporary workaround till it gets fixed?

**Comment by anntzer** [comment]
Re: the CoreFoundation warning: can you try explicitly setting it to Agg (`matplotlib.use("agg")` at the very top **before importing pyplot**) first?
backend_interagg seems to come with pycharm, not sure if anything they do interacts badly with anything else :/

**Comment by Yongzhi-GenapSys** [comment]
Hi marcotama，have you ever seen 'Could not convert glyph to bitmap' error in your case? Thanks.

**Comment by marcotama** [comment]
Hi, @Yongzhi-GenapSys , I don't think I have seen that one.

**Comment by anntzer** [comment]
Actually I think I found a fix (on Py3.7): you can add
```
os.register_at_fork(after_in_child=lambda: _get_font.cache_clear())
```
at the toplevel of font_manager.py (protected by an `if hasattr(os, 'register_at_fork')` so that's only for Py3.7+/unix).

Basically, the cached FT2Font objects holds a reference to a FT_Library, which becomes invalid after the fork() (or perhaps multiple processes start overwriting each other's FT_Library, I don't know).  So we just drop the cache when forking.  (Technically, we can still fail if you fork at a really awkward time to cause a FT2Font object to be reused, but at least this should cover most "normal" cases...)

Can you confirm that this works?

**Comment by Yongzhi-GenapSys** [comment]
@anntzer, does this fix work for Py3.7 only or it also work for Py 2.7.12 ? Thanks.

**Comment by anntzer** [comment]
Py3.7 only

**Comment by marcotama** [comment]
That fixes the issue for me! Thanks!

**Comment by tedwards-flf** [comment]
Is there a good workaround for Python 2.7.16?

**Comment by tedwards-flf** [comment]
When I tried an older version of matplotlib (v1.5.3) then it worked. It doesn't work on v2.2.3

**Comment by astrofrog** [comment]
@tacaswell - I'm still seeing this issue on Python 3.6 but not 3.7. Is the plan to not fix this for 3.6?

**Comment by bostonautolytics** [comment]
just found this issue myself in 3.7.1.

I'm sending data analysis tasks to a Huey consumer and can't run multiple workers due to this issue.

Isolated problem to verify that it occurs with parallel calls of savefig()

**Comment by tacaswell** [comment]
@astrofrog and @bostonautolytics Does https://github.com/matplotlib/matplotlib/issues/13723#issuecomment-476125019 fix the problem?

**Comment by josiahjohnston** [comment]
FYI, I'm periodically getting this issue using matplotlib v3.3.3 in a FastAPI/Starlette web application that generates png figures. The webserver library makes extensive use of async calls, which could explain why the register_at_fork workaround described here (embodied in PR https://github.com/matplotlib/matplotlib/pull/14535) doesn't fix this issue. I'm not sure, and don't have time for deeper digging tonight.

For the moment, I'm falling back to the hacky try/retry/gracefully give up approach described in a different thread I encountered today.

**Comment by jklymak** [comment]
Probably a good idea to open a new issue if its not working with current matplotlib....

**Comment by matandro** [comment]
We just upgraded from 3.2.1 to 3.3.4 and had the issue. Went back to 3.2.1 -> problem solved. So its something in the middle there (python 3.7.9)

**Comment by wentasah** [comment]
As mentioned above, the issue is still present in 3.3.4. However, it seems that it's gone for 3.4.0 and later. For 3.5.0b1, there is a fix #19618 which seems to address the same problem.

For 3.3.4, I worked around the problem by wrapping every call to `fig.savefig` with a lock. For example:

```python
import threading
plot_lock = threading.Lock()

with plot_lock:
    fig.savefig(b, format="svg", ...)
    plt.close(fig)
```

**Comment by tacaswell** [comment]
@wentasah The fix in #19618 is for threading, where as this issue is with multi-process.
